DeepSpeech.py

```diff
@@ -1710,56 +706,340 @@ def export():
                                       restore_op_name, filename_tensor_name,
                                       output_graph_path, clear_devices, '')
 
-            log_info('Models exported at %s' % (FLAGS.export_dir))
+            log.info('Models exported at %s' % (FLAGS.export_dir))
         except RuntimeError:
-            log_error(sys.exc_info()[1])
+            log.error(sys.exc_info()[1])
 
+class ClusterEndPoint(ClusterMessagingClient):
+    '''
+    Local cluster messaging end point.
+    '''
+    def __init__(self, cluster_spec, task):
+        ClusterMessagingClient.__init__(self, cluster_spec, 'worker', task)
+        self._index_lock = Lock()
+        self._get_gpus_per_worker_lock = Lock()
+        self._index = 0
+        self._event = Event()
+        # Public member to set an event callback after the model feeder got created.
+        # Reason for not being a ctor parameter: Creating the model feeder requires
+        # the cluster GPU configuration that in turn already requires messaging to
+        # query the GPU configuration of each worker.
+        self.on_start_data_set = None
+        self.gpus_per_worker = []
+
+    def _start_data_set(self, data_set_index):
+        if self.on_start_data_set:
+            self.on_start_data_set(data_set_index)
+
+    def start_data_set(self, data_set_index, index_offset=0):
+        '''
+        Starts a data_set on every node of the cluster.
+        Should only be called by chief worker.
```

Comment:
Is it possible to enforce this? RPC would reject a request done to a non-chief worker ?

Reply:
This is a design instruction/hint - all of this code is anyhow internal, one-file and 100% self-owned. Would be like controlling yourself.